File: app/src/main/java/com/aegamesi/steamtrade/fragments/FragmentProfile.java
package com.aegamesi.steamtrade.fragments;

import android.annotation.SuppressLint;
import android.content.DialogInterface;
import android.content.Intent;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AlertDialog;
import android.text.Html;
import android.text.method.LinkMovementMethod;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import com.aegamesi.steamtrade.R;
import com.aegamesi.steamtrade.lib.android.PicassoImageGetter;
import com.aegamesi.steamtrade.steam.SteamService;
import com.aegamesi.steamtrade.steam.SteamUtil;
import com.aegamesi.steamtrade.steam.SteamWeb;
import com.squareup.picasso.Picasso;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.Locale;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import de.hdodenhof.circleimageview.CircleImageView;
import uk.co.thomasc.steamkit.base.generated.steamlanguage.EFriendRelationship;
import uk.co.thomasc.steamkit.base.generated.steamlanguage.EPersonaState;
import uk.co.thomasc.steamkit.base.generated.steamlanguage.EResult;
import uk.co.thomasc.steamkit.steam3.handlers.steamfriends.callbacks.IgnoreFriendCallback;
import uk.co.thomasc.steamkit.steam3.handlers.steamfriends.callbacks.PersonaStateCallback;
import uk.co.thomasc.steamkit.steam3.handlers.steamfriends.callbacks.ProfileInfoCallback;
import uk.co.thomasc.steamkit.steam3.handlers.steamfriends.callbacks.SteamLevelCallback;
import uk.co.thomasc.steamkit.steam3.steamclient.callbackmgr.CallbackMsg;
import uk.co.thomasc.steamkit.types.steamid.SteamID;
import uk.co.thomasc.steamkit.util.cSharp.events.ActionT;

public class FragmentProfile extends FragmentBase implements View.OnClickListener {
	public SteamID id;
	public int steam_level = -1;
	public EFriendRelationship relationship;
	public EPersonaState state;
	public String name;
	public String game;
	public String avatar;

	public CircleImageView avatarView;
	public TextView nameView;
	public TextView statusView;
	public TextView summaryView;
	public TextView levelView;
	public Button chatButton;
	public Button addFriendButton;
	public Button removeFriendButton;
	public Button blockFriendButton;
	public Button unblockFriendButton;
	public Button viewSteamButton;
	public Button viewSteamRepButton;
	public Button viewLibraryButton;

	public ProfileInfoCallback profile_info = null;
	public PersonaStateCallback persona_info = null;

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		if (abort)
			return;

		if (getArguments().containsKey("steamId")) {
			id = new SteamID(getArguments().getLong("steamId"));
		} else {
			String url = getArguments().getString("url");

			if (url != null) {
				Matcher matcher = Pattern.compile("steamcommunity.com/id/([a-zA-Z0-9]+)").matcher(url);
				if (matcher.find()) {
					id = null;
					String vanity = matcher.group(1);
					(new ResolveVanityURLTask()).execute(vanity);
				} else {
					matcher = Pattern.compile("steamcommunity.com/profiles/([0-9]+)").matcher(url);
					if (matcher.find())
						id = new SteamID(Long.parseLong(matcher.group(1)));
					else
						id = null;
				}
			}
		}

	}

	@Override
	public void handleSteamMessage(CallbackMsg msg) {
		msg.handle(SteamLevelCallback.class, new ActionT<SteamLevelCallback>() {
			@Override
			public void call(SteamLevelCallback obj) {
				if (id != null && obj.getLevelMap().containsKey(id)) {
					steam_level = obj.getLevelMap().get(id);
					updateView();
				}
			}
		});
		msg.handle(ProfileInfoCallback.class, new ActionT<ProfileInfoCallback>() {
			@Override
			public void call(ProfileInfoCallback obj) {
				profile_info = obj;
				updateView();
			}
		});
		msg.handle(PersonaStateCallback.class, new ActionT<PersonaStateCallback>() {
			@Override
			public void call(PersonaStateCallback obj) {
				if (id != null && id.equals(obj.getFriendID())) {
					persona_info = obj;
					updateView();
				}
			}
		});
		msg.handle(IgnoreFriendCallback.class, new ActionT<IgnoreFriendCallback>() {
			@Override
			public void call(IgnoreFriendCallback obj) {
				boolean success = obj.getResult() == EResult.OK;
				int stringResource = success ? R.string.action_successful : R.string.action_failed;
				Toast.makeText(activity(), stringResource, Toast.LENGTH_SHORT).show();
				updateView();
			}
		});
	}

	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
		inflater = activity().getLayoutInflater();
		View view = inflater.inflate(R.layout.fragment_profile, container, false);

		avatarView = view.findViewById(R.id.profile_avatar);
		nameView = view.findViewById(R.id.profile_name);
		statusView = view.findViewById(R.id.profile_status);
		summaryView = view.findViewById(R.id.profile_summary);
		levelView = view.findViewById(R.id.profile_level);
		chatButton = view.findViewById(R.id.profile_button_chat);
		removeFriendButton = view.findViewById(R.id.profile_button_remove_friend);
		addFriendButton = view.findViewById(R.id.profile_button_add_friend);
		viewSteamButton = view.findViewById(R.id.profile_button_viewsteam);
		viewSteamRepButton = view.findViewById(R.id.profile_button_viewsteamrep);
		blockFriendButton = view.findViewById(R.id.profile_button_block_friend);
		unblockFriendButton = view.findViewById(R.id.profile_button_unblock_friend);
		viewLibraryButton = view.findViewById(R.id.profile_button_library);

		chatButton.setOnClickListener(this);
		removeFriendButton.setOnClickListener(this);
		addFriendButton.setOnClickListener(this);
		viewSteamButton.setOnClickListener(this);
		viewSteamRepButton.setOnClickListener(this);
		blockFriendButton.setOnClickListener(this);
		unblockFriendButton.setOnClickListener(this);
		viewLibraryButton.setOnClickListener(this);

		nameView.setSelected(true);
		statusView.setSelected(true);
		summaryView.setMovementMethod(LinkMovementMethod.getInstance());

		updateView();
		return view;
	}

	@Override
	public void onStart() {
		super.onStart();

		requestInfo();
		// fetch the data
		//(new ProfileFetchTask()).execute();
	}

	public void updateView() {
		if (activity() == null || activity().steamFriends == null)
			return;

		relationship = activity().steamFriends.getFriendRelationship(id);
		if (relationship == EFriendRelationship.Friend || persona_info == null) {
			state = activity().steamFriends.getFriendPersonaState(id);
			relationship = activity().steamFriends.getFriendRelationship(id);
			name = activity().steamFriends.getFriendPersonaName(id);
			game = activity().steamFriends.getFriendGamePlayedName(id);
			avatar = SteamUtil.bytesToHex(activity().steamFriends.getFriendAvatar(id)).toLowerCase(Locale.US);
		} else {
			// use the found persona info stuff
			state = persona_info.getState();
			relationship = activity().steamFriends.getFriendRelationship(id);
			name = persona_info.getName();
			game = persona_info.getGameName();
			avatar = SteamUtil.bytesToHex(persona_info.getAvatarHash()).toLowerCase(Locale.US);
		}

		addFriendButton.setText((relationship == EFriendRelationship.RequestRecipient) ? R.string.friend_accept : R.string.friend_add);
		unblockFriendButton.setText((relationship == EFriendRelationship.RequestRecipient) ? R.string.friend_ignore : R.string.friend_unblock);

		if (profile_info != null) {
			String summary_raw = profile_info.getSummary();
			String summary = SteamUtil.parseBBCode(summary_raw);
			Html.ImageGetter imageGetter = new PicassoImageGetter(summaryView, summaryView.getContext());
			summaryView.setText(Html.fromHtml(summary, Html.FROM_HTML_MODE_LEGACY, imageGetter, null));
			summaryView.setMovementMethod(new LinkMovementMethod());
		}

		if (steam_level == -1) {
			levelView.setText(R.string.unknown);
		} else {
			levelView.setText(String.valueOf(steam_level));
		}

		setTitle(name);
		nameView.setText(name);

		//Profile loader for friend/user.
		//avatarView.setImageResource(R.drawable.default_avatar);
		if (avatar != null && avatar.length() == 40 && !avatar.equals("0000000000000000000000000000000000000000")) {
			Picasso.with(getContext())
					.load("http://media.steampowered.com/steamcommunity/public/images/avatars/" + avatar.substring(0, 2) + "/" + avatar + "_full.jpg")
					.into(avatarView);
		}

		if (game != null && game.length() > 0)
			statusView.setText(String.format(getString(R.string.profile_playing_game), game));
		else
			statusView.setText(state.toString());

		int color = ContextCompat.getColor(getContext(), R.color.steam_online);
		if (relationship == EFriendRelationship.Blocked || relationship == EFriendRelationship.Ignored || relationship == EFriendRelationship.IgnoredFriend)
			color = ContextCompat.getColor(getContext(), R.color.steam_blocked);
		else if (game != null && game.length() > 0)
			color = ContextCompat.getColor(getContext(), R.color.steam_game);
		else if (state == EPersonaState.Offline || state == null)
			color = ContextCompat.getColor(getContext(), R.color.steam_offline);

		nameView.setTextColor(color);
		statusView.setTextColor(color);
		avatarView.setBorderColor(color);

		// things to do if we are not friends
		boolean isFriend = relationship == EFriendRelationship.Friend || relationship == EFriendRelationship.IgnoredFriend;
		boolean isSelf = SteamService.singleton.steamClient.getSteamId().equals(id);
		boolean isBlocked = relationship == EFriendRelationship.Blocked || relationship == EFriendRelationship.Ignored || relationship == EFriendRelationship.IgnoredFriend;
		boolean isFriendRequest = relationship == EFriendRelationship.RequestRecipient;

		if (!isFriend) {
			statusView.setText(relationship.toString());
			addFriendButton.setEnabled(id != null);
		}

		//Fix relationship if you view yourself.
		if (relationship == EFriendRelationship.None && isSelf )
			statusView.setText(state.toString());

		// visibility of buttons and stuff
		addFriendButton.setVisibility((!isFriend && !isSelf && !isBlocked) ? View.VISIBLE : View.GONE);
		removeFriendButton.setVisibility((isFriend && !isSelf) ? View.VISIBLE : View.GONE);
		chatButton.setVisibility((isFriend && !isSelf && !isBlocked) ? View.VISIBLE : View.GONE);
		blockFriendButton.setVisibility((!isBlocked && !isSelf) ? View.VISIBLE : View.GONE);
		unblockFriendButton.setVisibility((isFriendRequest || isBlocked) ? View.VISIBLE : View.GONE);
	}

	private void requestInfo() {
		if (id != null) {
			if (profile_info == null) {
				activity().steamFriends.requestProfileInfo(id);
			}

			int request_flags = 1 | 2 | 4 | 8 | 16 | 32 | 64 | 128 | 256 | 512 | 1024;
			relationship = activity().steamFriends.getFriendRelationship(id);
			if (relationship != EFriendRelationship.Friend && persona_info == null) {
				SteamID myId = SteamService.singleton.steamClient.getSteamId();
				if (!id.equals(myId)) {
					activity().steamFriends.requestFriendInfo(id, request_flags);
				}
			}

			if (steam_level == -1) {
				activity().steamFriends.requestSteamLevel(id);
			}
		}
	}

	@Override
	public void onClick(View view) {
		if (view == removeFriendButton) {
			AlertDialog.Builder builder = new AlertDialog.Builder(activity());
			builder.setMessage(String.format(getString(R.string.friend_remove_message), activity().steamFriends.getFriendPersonaName(id)));
			builder.setTitle(R.string.friend_remove);
			builder.setPositiveButton(R.string.ok, new DialogInterface.OnClickListener() {
				public void onClick(DialogInterface dialog, int id) {
					activity().steamFriends.removeFriend(FragmentProfile.this.id);
					Toast.makeText(activity(), String.format(getString(R.string.friend_removed), activity().steamFriends.getFriendPersonaName(FragmentProfile.this.id)), Toast.LENGTH_LONG).show();
					activity().browseToFragment(new FragmentFriends(), true);
				}
			});
			builder.setNegativeButton(R.string.cancel, new DialogInterface.OnClickListener() {
				public void onClick(DialogInterface dialog, int id) {
				}
			});
			builder.create().show();
		}
		if (view == addFriendButton) {
			activity().steamFriends.addFriend(id);
			Toast.makeText(activity(), R.string.friend_add_success, Toast.LENGTH_LONG).show();
		}
		if (view == chatButton) {
			Fragment fragment = new FragmentChat();
			Bundle bundle = new Bundle();
			bundle.putLong("steamId", id.convertToLong());
			bundle.putBoolean("fromProfile", true);
			fragment.setArguments(bundle);
			activity().browseToFragment(fragment, true);
		}
		if (view == viewSteamButton) {
			String url = "http://steamcommunity.com/profiles/" + id.convertToLong();
			FragmentWeb.openPage(activity(), url, false);
		}
		if (view == viewSteamRepButton) {
			String steamRepUrl = "http://steamrep.com/profiles/" + id.convertToLong() + "/";
			Intent browse = new Intent(Intent.ACTION_VIEW, Uri.parse(steamRepUrl));
			startActivity(browse);
		}
		if (view == blockFriendButton) {
			activity().steamFriends.ignoreFriend(id, true);
		}
		if (view == unblockFriendButton) {
			activity().steamFriends.ignoreFriend(id, false);
		}
		if (view == viewLibraryButton) {
			Fragment fragment = new FragmentLibrary();
			Bundle bundle = new Bundle();
			bundle.putLong("id", id.convertToLong());
			fragment.setArguments(bundle);
			activity().browseToFragment(fragment, true);
		}
	}

	@SuppressLint("StaticFieldLeak")
	private class ResolveVanityURLTask extends AsyncTask<String, Void, SteamID> {
		@Override
		protected SteamID doInBackground(String... args) {
			String vanity = args[0];

			String api_url = "https://api.steampowered.com/ISteamUser/ResolveVanityURL/v1/?key=" + SteamUtil.webApiKey + "&format=json&vanityurl=" + vanity;
			String response = SteamWeb.fetch(api_url, "GET", null, "");
			if (response.length() == 0)
				return null;
			JSONObject response_obj;
			try {
				response_obj = new JSONObject(response);
				response_obj = response_obj.getJSONObject("response");
				int result = response_obj.getInt("success");
				if (result != 1)
					return null;

				String steam_id = response_obj.getString("steamid");
				long long_value = Long.parseLong(steam_id);
				return new SteamID(long_value);
			} catch (JSONException e) {
				e.printStackTrace();
				return null;
			}
		}

		@Override
		protected void onPostExecute(SteamID id) {
			FragmentProfile.this.id = id;
			requestInfo();
			updateView();
		}
	}
}

File: app/src/main/java/com/aegamesi/steamtrade/fragments/FragmentProfile.kt
package com.aegamesi.steamtrade.fragments

import android.annotation.SuppressLint
import android.content.Intent
import android.net.Uri
import android.os.AsyncTask
import android.os.Bundle
import android.text.Html
import android.text.method.LinkMovementMethod
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.core.content.ContextCompat
import com.aegamesi.steamtrade.R
import com.aegamesi.steamtrade.libs.GlideImageGetter
import com.aegamesi.steamtrade.steam.SteamService
import com.aegamesi.steamtrade.steam.SteamUtil
import com.aegamesi.steamtrade.steam.SteamWeb
import com.bumptech.glide.Glide
import kotlinx.android.synthetic.main.fragment_profile.*
import org.json.JSONException
import org.json.JSONObject
import uk.co.thomasc.steamkit.base.generated.steamlanguage.EFriendRelationship
import uk.co.thomasc.steamkit.base.generated.steamlanguage.EPersonaState
import uk.co.thomasc.steamkit.base.generated.steamlanguage.EResult
import uk.co.thomasc.steamkit.steam3.handlers.steamfriends.callbacks.IgnoreFriendCallback
import uk.co.thomasc.steamkit.steam3.handlers.steamfriends.callbacks.PersonaStateCallback
import uk.co.thomasc.steamkit.steam3.handlers.steamfriends.callbacks.ProfileInfoCallback
import uk.co.thomasc.steamkit.steam3.handlers.steamfriends.callbacks.SteamLevelCallback
import uk.co.thomasc.steamkit.steam3.steamclient.callbackmgr.CallbackMsg
import uk.co.thomasc.steamkit.types.steamid.SteamID
import uk.co.thomasc.steamkit.util.cSharp.events.ActionT
import java.util.*
import java.util.regex.Pattern

class FragmentProfile : FragmentBase(), View.OnClickListener {
    var id: SteamID? = null
    var steamLevel = -1
    private lateinit var relationship: EFriendRelationship
    private var state: EPersonaState? = null
    lateinit var name: String
    private var game: String? = null
    private var avatar: String? = null
    
    var profileInfo: ProfileInfoCallback? = null
    var personaInfo: PersonaStateCallback? = null

    companion object {
        private const val AVATAR_BASE_URL = "http://media.steampowered.com/steamcommunity/public/images/avatars/"
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        if (abort) return

        Log.i("FragmentProfile", "created")

        if (arguments!!.containsKey("steamId")) {
            id = SteamID(arguments!!.getLong("steamId"))
        } else {
            val url = arguments!!.getString("url")

            if (url != null) {
                var matcher = Pattern.compile("steamcommunity.com/id/([a-zA-Z0-9]+)").matcher(url)
                if (matcher.find()) {
                    id = null
                    val vanity = matcher.group(1)
                    ResolveVanityURLTask().execute(vanity)
                } else {
                    matcher = Pattern.compile("steamcommunity.com/profiles/([0-9]+)").matcher(url)
                    if (matcher.find())
                        id = SteamID(java.lang.Long.parseLong(matcher.group(1)))
                    else
                        id = null
                }
            }
        }

    }

    override fun handleSteamMessage(msg: CallbackMsg) {
        msg.handle(SteamLevelCallback::class.java, object : ActionT<SteamLevelCallback>() {
            override fun call(obj: SteamLevelCallback) {
                if (id != null && obj.levelMap.containsKey(id)) {
                    steamLevel = obj.levelMap[id]!!
                    updateView()
                }
            }
        })
        msg.handle(ProfileInfoCallback::class.java, object : ActionT<ProfileInfoCallback>() {
            override fun call(obj: ProfileInfoCallback) {
                profileInfo = obj
                updateView()
            }
        })
        msg.handle(PersonaStateCallback::class.java, object : ActionT<PersonaStateCallback>() {
            override fun call(obj: PersonaStateCallback) {
                if (id != null && id == obj.friendID) {
                    personaInfo = obj
                    updateView()
                }
            }
        })
        msg.handle(IgnoreFriendCallback::class.java, object : ActionT<IgnoreFriendCallback>() {
            override fun call(obj: IgnoreFriendCallback) {
                val success = obj.result == EResult.OK
                val stringResource = if (success) R.string.action_successful else R.string.action_failed
                Toast.makeText(activity(), stringResource, Toast.LENGTH_SHORT).show()
                updateView()
            }
        })
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_profile, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        profile_button_chat.setOnClickListener(this)
        profile_button_remove_friend.setOnClickListener(this)
        profile_button_add_friend.setOnClickListener(this)
        profile_button_viewsteam.setOnClickListener(this)
        profile_button_viewsteamrep.setOnClickListener(this)
        profile_button_block_friend.setOnClickListener(this)
        profile_button_unblock_friend.setOnClickListener(this)
        profile_button_library.setOnClickListener(this)

        profile_name.isSelected = true
        profile_status.isSelected = true
        profile_summary.movementMethod = LinkMovementMethod.getInstance()

        updateView()
    }

    override fun onStart() {
        super.onStart()
        requestInfo()
    }

    fun updateView() {
        if (activity() == null)
            return

        relationship = activity()!!.steamFriends.getFriendRelationship(id)
        if (relationship == EFriendRelationship.Friend || personaInfo == null) {
            state = activity()!!.steamFriends.getFriendPersonaState(id)
            relationship = activity()!!.steamFriends.getFriendRelationship(id)
            name = activity()!!.steamFriends.getFriendPersonaName(id)
            game = activity()!!.steamFriends.getFriendGamePlayedName(id)
            avatar = SteamUtil.bytesToHex(activity()!!.steamFriends.getFriendAvatar(id)).toLowerCase(Locale.US)
        } else {
            // use the found persona info stuff
            state = personaInfo!!.state
            relationship = activity()!!.steamFriends.getFriendRelationship(id)
            name = personaInfo!!.name
            game = personaInfo!!.gameName
            avatar = SteamUtil.bytesToHex(personaInfo!!.avatarHash).toLowerCase(Locale.US)
        }

        profile_button_add_friend.setText(if (relationship == EFriendRelationship.RequestRecipient) R.string.friend_accept else R.string.friend_add)
        profile_button_unblock_friend.setText(if (relationship == EFriendRelationship.RequestRecipient) R.string.friend_ignore else R.string.friend_unblock)

        if (profileInfo != null) {
            val summaryRaw = profileInfo!!.summary
            val summary = SteamUtil.parseBBCode(summaryRaw)
            val imageGetter = GlideImageGetter(profile_summary, profile_summary.context)
            profile_summary.text = Html.fromHtml(summary, Html.FROM_HTML_MODE_LEGACY, imageGetter, null)
            profile_summary.movementMethod = LinkMovementMethod()
        }

        if (steamLevel == -1) {
            profile_level.setText(R.string.unknown)
        } else {
            profile_level.text = steamLevel.toString()
        }

        setTitle(name)
        profile_name.text = name

        Glide.with(activity()!!.applicationContext)
                .load(AVATAR_BASE_URL + avatar!!.substring(0, 2) + "/" + avatar + "_full.jpg")
                .placeholder(R.drawable.default_avatar)
                .error(R.drawable.default_avatar)
                .into(profile_avatar)

        if (game!!.isEmpty())
            profile_status.text = String.format(getString(R.string.profile_playing_game), game)
        else
            profile_status.text = state!!.toString()

        var color = ContextCompat.getColor(activity()!!.applicationContext, R.color.steam_online)
        if (relationship == EFriendRelationship.Blocked || relationship == EFriendRelationship.Ignored || relationship == EFriendRelationship.IgnoredFriend)
            color = ContextCompat.getColor(activity()!!.applicationContext, R.color.steam_blocked)
        else if (game != null && game!!.isNotEmpty())
            color = ContextCompat.getColor(activity()!!.applicationContext, R.color.steam_game)
        else if (state == EPersonaState.Offline || state == null)
            color = ContextCompat.getColor(activity()!!.applicationContext, R.color.steam_offline)

        profile_name.setTextColor(color)
        profile_status.setTextColor(color)
        profile_avatar.borderColor = color

        // things to do if we are not friends
        val isFriend = relationship == EFriendRelationship.Friend || relationship == EFriendRelationship.IgnoredFriend
        val isSelf = SteamService.singleton!!.steamClient!!.steamId == id
        val isBlocked = relationship == EFriendRelationship.Blocked || relationship == EFriendRelationship.Ignored || relationship == EFriendRelationship.IgnoredFriend
        val isFriendRequest = relationship == EFriendRelationship.RequestRecipient

        if (!isFriend) {
            profile_status.text = relationship.toString()
            profile_button_add_friend.isEnabled = id != null
        }

        //Fix relationship if you view yourself.
        if (relationship == EFriendRelationship.None && isSelf)
            profile_status.text = state!!.toString()

        // visibility of buttons and stuff
        profile_button_add_friend.visibility = if (!isFriend && !isSelf && !isBlocked) View.VISIBLE else View.GONE
        profile_button_remove_friend.visibility = if (isFriend && !isSelf) View.VISIBLE else View.GONE
        profile_button_chat.visibility = if (isFriend && !isSelf && !isBlocked) View.VISIBLE else View.GONE
        profile_button_block_friend.visibility = if (!isBlocked && !isSelf) View.VISIBLE else View.GONE
        profile_button_unblock_friend.visibility = if (isFriendRequest || isBlocked) View.VISIBLE else View.GONE
    }

    private fun requestInfo() {
        if (id != null) {
            if (profileInfo == null) {
                activity()!!.steamFriends.requestProfileInfo(id!!)
            }

            val requestFlags = 1 or 2 or 4 or 8 or 16 or 32 or 64 or 128 or 256 or 512 or 1024
            relationship = activity()!!.steamFriends.getFriendRelationship(id)
            if (relationship != EFriendRelationship.Friend && personaInfo == null) {
                val myId = SteamService.singleton!!.steamClient!!.steamId
                if (id != myId) {
                    activity()!!.steamFriends.requestFriendInfo(id, requestFlags)
                }
            }

            if (steamLevel == -1) {
                activity()!!.steamFriends.requestSteamLevel(id)
            }
        }
    }

    override fun onClick(view: View) {
        if (view == profile_button_remove_friend) {
            val builder = AlertDialog.Builder(activity()!!)
            builder.setMessage(String.format(getString(R.string.friend_remove_message), activity()!!.steamFriends.getFriendPersonaName(id)))
            builder.setTitle(R.string.friend_remove)
            builder.setPositiveButton(android.R.string.ok) { _, _ ->
                activity()!!.steamFriends.removeFriend(this@FragmentProfile.id!!)
                Toast.makeText(activity(), String.format(getString(R.string.friend_removed), activity()!!.steamFriends.getFriendPersonaName(this@FragmentProfile.id)), Toast.LENGTH_LONG).show()
                activity()!!.browseToFragment(FragmentFriends(), true)
            }
            builder.setNegativeButton(android.R.string.cancel) { _, _ -> }
            builder.create().show()
        }
        if (view == profile_button_add_friend) {
            activity()!!.steamFriends.addFriend(id!!)
            Toast.makeText(activity(), R.string.friend_add_success, Toast.LENGTH_LONG).show()
        }
        if (view == profile_button_chat) {
            val fragment = FragmentChat()
            val bundle = Bundle()
            bundle.putLong("steamId", id!!.convertToLong())
            bundle.putBoolean("fromProfile", true)
            fragment.arguments = bundle
            activity()!!.browseToFragment(fragment, true)
        }
        if (view == profile_button_viewsteam) {
            val url = "http://steamcommunity.com/profiles/" + id!!.convertToLong()
            FragmentWeb.openPage(activity()!!, url, false)
        }
        if (view == profile_button_viewsteamrep) {
            val steamRepUrl = "http://steamrep.com/profiles/" + id!!.convertToLong() + "/"
            val browse = Intent(Intent.ACTION_VIEW, Uri.parse(steamRepUrl))
            startActivity(browse)
        }
        if (view == profile_button_block_friend) {
            activity()!!.steamFriends.ignoreFriend(id, true)
        }
        if (view == profile_button_unblock_friend) {
            activity()!!.steamFriends.ignoreFriend(id, false)
        }
        if (view == profile_button_library) {
            val fragment = FragmentLibrary()
            val bundle = Bundle()
            bundle.putLong("id", id!!.convertToLong())
            fragment.arguments = bundle
            activity()!!.browseToFragment(fragment, true)
        }
    }

    @SuppressLint("StaticFieldLeak")
    private inner class ResolveVanityURLTask : AsyncTask<String, Void, SteamID>() {
        override fun doInBackground(vararg args: String): SteamID? {
            val vanity = args[0]

            val apiUrl = "https://api.steampowered.com/ISteamUser/ResolveVanityURL/v1/?key=" + SteamUtil.webApiKey + "&format=json&vanityurl=" + vanity
            val response = SteamWeb.fetch(apiUrl, "GET", null, "")
            if (response.isEmpty())
                return null
            var responseObj: JSONObject
            try {
                responseObj = JSONObject(response)
                responseObj = responseObj.getJSONObject("response")
                val result = responseObj.getInt("success")
                if (result != 1)
                    return null

                val steamID = responseObj.getString("steamID")
                val longValue = java.lang.Long.parseLong(steamID)
                return SteamID(longValue)
            } catch (e: JSONException) {
                e.printStackTrace()
                return null
            }

        }

        override fun onPostExecute(id: SteamID) {
            this@FragmentProfile.id = id
            requestInfo()
            updateView()
        }
    }
}